examples/fabric-bridge-app/linux/DeviceManager.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "DeviceManager.h"
+
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/EventLogging.h>
+#include <app/reporting/reporting.h>
+#include <app/server/Server.h>
+#include <app/util/af-types.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/endpoint-config-api.h>
+#include <app/util/util.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/ZclString.h>
+
+#include <cstdio>
+#include <string>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::Credentials;
+using namespace chip::Inet;
+using namespace chip::Transport;
+using namespace chip::DeviceLayer;
+using namespace chip::app::Clusters;
+
+namespace {
+constexpr uint8_t kMaxRetries = 10;
+} // namespace
+
+DeviceManager::DeviceManager()
+{
+    memset(mDevices, 0, sizeof(mDevices));
+    mFirstDynamicEndpointId = static_cast<chip::EndpointId>(
+        static_cast<int>(emberAfEndpointFromIndex(static_cast<uint16_t>(emberAfFixedEndpointCount() - 1))) + 1);
+    mCurrentEndpointId = mFirstDynamicEndpointId;
+}
+
+int DeviceManager::AddDeviceEndpoint(Device * dev, EmberAfEndpointType * ep,
+                                     const chip::Span<const EmberAfDeviceType> & deviceTypeList,
+                                     const chip::Span<chip::DataVersion> & dataVersionStorage, chip::EndpointId parentEndpointId)
+{
+    uint8_t index = 0;
+    while (index < CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT)
+    {
+        if (nullptr == mDevices[index])
+        {
+            mDevices[index] = dev;
+            CHIP_ERROR err;
+            int retryCount = 0;
+            while (retryCount < kMaxRetries)
+            {
+                DeviceLayer::StackLock lock;
+                dev->SetEndpointId(mCurrentEndpointId);
+                dev->SetParentEndpointId(parentEndpointId);
+                err =
+                    emberAfSetDynamicEndpoint(index, mCurrentEndpointId, ep, dataVersionStorage, deviceTypeList, parentEndpointId);
```

Comment:
wouldn't be better to look for a free DynamicEndpointId and set it to that endpoint rather This retry loop?

Reply:
I agree, but we first need to implement a new ember function emberAfDynamicEndpointIdExists to check if a endpoint is free. At this moment, the only way to see if an endpoint is free is try to set it via  emberAfSetDynamicEndpoint and check the result.